### PR TITLE
fix: adjust router for insufficient asset

### PR DESCRIPTION
**Pull request by dmoka**
Since we introduced insufficient assets, the balance check won't work properly as we can have special case with selling fee payment asset for shitcoin, so we decided to remove it.

Furthermore we wanna disable route setting for insufficient asset

## TODO:
- [x] Rebenchmark pallet

### CODE REVIEW

File: integration-tests/src/dca.rs [hunk 1]
```diff
@@ -2853,6 +2854,8 @@ mod with_onchain_route {
 
 		TestNet::reset();
 		Hydra::execute_with(|| {
+			let ps = frame_system::Pallet::<Runtime>::providers(&AccountId32::from(ALICE));
```
Comment (mrq1911) [inline integration-tests/src/dca.rs:2857]: what those ps variables do?

File: pallets/route-executor/src/lib.rs [hunk 2]
```diff
@@ -310,6 +316,7 @@ pub mod pallet {
 				asset_in,
 				user_balance_of_asset_in_before_trade,
 				first_trade.amount_in,
+				existential_deposit, //TODO: add test for buy then fix it,
```
Comment (mrq1911) [inline pallets/route-executor/src/lib.rs:319]: TODO

File: pallets/route-executor/src/lib.rs [hunk 3]
```diff
@@ -454,16 +461,23 @@ impl<T: Config> Pallet<T> {
...
 
-			ensure!(
-				user_balance_of_asset_in_before_trade - spent_amount == user_balance_of_asset_in_after_trade,
-				Error::<T>::InvalidRouteExecution
-			);
+		let expected_user_balance_of_asset_in_after_trade = user_balance_of_asset_in_before_trade
+			.checked_sub(&spent_amount)
+			.ok_or(Error::<T>::InvalidRouteExecution)?;
+
+		if expected_user_balance_of_asset_in_after_trade < existential_deposit {
```
Comment (mrq1911) [inline pallets/route-executor/src/lib.rs:473]: this could be simply just part of `ensure!` predicate

**Review by mrq1911 — CHANGES_REQUESTED** [review]
pls bump versions

File: integration-tests/src/dca.rs (hunk 1 above)
Comment (dmoka) [reply integration-tests/src/dca.rs:2857]: it is still draft, about to removing them and clean the PR x)

**Comment by dmoka** [comment]
dont merge it yet, we will rework to get the ED from asset registry

File: pallets/route-executor/src/lib.rs [hunk 4]
```diff
@@ -454,16 +466,21 @@ impl<T: Config> Pallet<T> {
...
-				Error::<T>::InvalidRouteExecution
-			);
-		}
+		let expected_user_balance_of_asset_in_after_trade = user_balance_of_asset_in_before_trade
+			.checked_sub(&spent_amount)
+			.ok_or(Error::<T>::InvalidRouteExecution)?;
+
+		//If the user had leftover less than ED then it is wiped out, hence we can't check the balance precisely
+		ensure!(
+			expected_user_balance_of_asset_in_after_trade < existential_deposit
```
Comment (jak-pan) [inline pallets/route-executor/src/lib.rs:480]: We can also check if it's 0

Comment (dmoka) [reply pallets/route-executor/src/lib.rs:480]: You would check if the `user_balance_of_asset_in_after_trade` is equal 0?

Comment (dmoka) [reply pallets/route-executor/src/lib.rs:480]: what is the reason for checking it for 0?

**Review by mrq1911 — APPROVED** [review]
Rebench and LGTM

**Pull request merged**